File: src/main/java/io/github/intellij/dlanguage/project/DubListenerComponent.java
package io.github.intellij.dlanguage.project;

import com.intellij.openapi.module.Module;
import com.intellij.openapi.module.ModuleManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.startup.StartupActivity;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.openapi.vfs.impl.BulkVirtualFileListenerAdapter;
import org.jetbrains.annotations.NotNull;

import static com.intellij.openapi.vfs.VirtualFileManager.VFS_CHANGES;

/**
 * Created by francis on 1/27/2018.
 */
public class DubListenerComponent implements StartupActivity {

    @Override
    public void runActivity(@NotNull final Project project) {
        for (final Module module : ModuleManager.getInstance(project).getModules()) {
            final VirtualFile dubFile = DubConfigFileListener.getDubFileFromModule(module);
            if (dubFile != null) {
                project
                    .getMessageBus()
                    .connect()
                    .subscribe(VFS_CHANGES, new BulkVirtualFileListenerAdapter(new DubConfigFileListener(dubFile, project, module)));
            }
        }
    }
}

File: dub/src/main/kotlin/io/github/intellij/dub/project/DubListenerComponent.kt
package io.github.intellij.dub.project

import com.intellij.openapi.module.ModuleManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.startup.ProjectActivity
import com.intellij.openapi.startup.StartupActivity
import com.intellij.openapi.vfs.VirtualFileManager
import com.intellij.openapi.vfs.impl.BulkVirtualFileListenerAdapter
import io.github.intellij.dub.project.DubConfigFileListener.Companion.getDubFileFromModule

/**
 * Created by francis on 1/27/2018.
 */
class DubListenerComponent : ProjectActivity {
    override suspend fun execute(project: Project) {
        for (module in ModuleManager.getInstance(project).modules) {
            val dubFile = getDubFileFromModule(module!!)
            if (dubFile != null) {
                project
                    .messageBus
                    .connect()
                    .subscribe(
                        VirtualFileManager.VFS_CHANGES,
                        BulkVirtualFileListenerAdapter(DubConfigFileListener(dubFile, project, module))
                    )
            }
        }
    }
}
